scripts/code_style_rewrite_branch.py

```diff
@@ -0,0 +1,214 @@
+    if result.returncode != 0:
+        print("Error detaching HEAD.", file=sys.stderr)
+        return False
+    result = subprocess.run(["git", "branch", "-D", branch_name], check=False)
+    if result.returncode != 0:
+        print("Error deleting branch '" + branch_name + "'.", file=sys.stderr)
+        return False
+    return True
+
+def rebase_on(branch: str, commit: str, new_branch: str) -> bool:
+    """
+    Make a new branch with the given name that consists of the given branch
+    rebased on the given commit.
+    """
+    # Create a new branch from the existing branch
+    result = subprocess.run(["git", "checkout", branch], check=False)
+    if result.returncode != 0:
+        print("Error checking out branch '"+branch+"'.", file=sys.stderr)
+        return False
+    result = subprocess.run(["git", "checkout", "-b", new_branch], check=False)
+    if result.returncode != 0:
+        print("Error creating rebase destination branch '"+new_branch+"'.", \
+                file=sys.stderr)
+        return False
+
+    # Rebase the branch atop the given commit
+    result = subprocess.run(["git", "rebase", commit], check=False)
+    if result.returncode != 0:
+        print("Error rebasing branch '"+branch+"' on commit '"+commit+"'.", \
+                file=sys.stderr)
+        delete_branch(new_branch)
+        return False
+
+    return True
+
+def restyle_commit_onto_current(commit_hash: str) -> bool:
+    """
+    Re-style the given commit and add the result as a new commit on the given
+    branch.
+    """
+    # Get a list of changed files in the commit
+    result = subprocess.run(["git", "diff", "--name-only", commit_hash, \
+            commit_hash + "~"], stdout=subprocess.PIPE, check=False)
+    if result.returncode != 0:
+        print("Error getting changed files for commit " + commit_hash + \
+                ".", file=sys.stderr)
+        return False
+    changed_files = str(result.stdout, "ascii").strip().split()
+
+    # Checkout changed files to their state in the old commit
+    for changed_file in changed_files:
+        result = subprocess.run(["git", "checkout", commit_hash, "--", \
+                changed_file], check=False)
+        if result.returncode != 0:
+            print("Error checking out changed file '" + changed_file + \
+                    "' to commit " + commit_hash + ".", file=sys.stderr)
+            return False
+
+    # Restyle the changed files to the new style
+    if code_style.fix_style(changed_files) != 0:
+        return False
+
+    # Add the changed files
+    for changed_file in changed_files:
+        result = subprocess.run(["git", "add", changed_file], check=False)
+        if result.returncode != 0:
+            print("Error adding changed file '" + changed_file + "'.", \
+                    file=sys.stderr)
+            return False
+
+    # Commit the newly added files
+    result = subprocess.run(["git", "commit", "--reuse-message=" + \
```

Comment:
Ultimately: this could be an empty commit if the original was just a style fix.

Reply:
It can't because we're also recreating the original changes (rather than adding a separate commit for style correction). This commit can only be empty if the original was also empty.

Reply:
Oh actually no you're right. I'll have to think about how to address this edge case.

Reply:
A work-around, perhaps not fully satisfying, would be for users to manually remove style-only commits (with `rebase -i` or whatever) before using this script. But for this to work this script should at least be able to give an indication that it's what's happening - probably by looking for `nothing added to commit` in the output of `git commit`.

Reply:
We can't reasonably ask users to detect, specifically, style-only commits _which would become no-ops after restyling_.

Reply:
I think we can use:
```bash
git diff --cached --exit-code
```
To check if there is anything staged for commit. If not, we can just skip that commit with some message.

Reply:
In https://github.com/Mbed-TLS/mbedtls-docs/pull/71, for now, I've chosen to allow empty commits. This preserves empty commits if the author originally had one for some reason, and also preserves style-only commits that become empty. The advantage of keeping empty commits is that we keep their commit message, which might be wanted. The downside is that I think that the most likely case is that the commit is fluff with a boring commit message. But it's hard to tell because it's a rare case overall. On the gripping hand, allowing empty commits has the advantage of mechanical simplicity: there are fewer conditions to manage, and fewer things that can go wrong.

I'm not sure that's the right choice. We can revisit it.